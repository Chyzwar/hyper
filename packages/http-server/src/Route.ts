/* eslint-disable @typescript-eslint/no-misused-promises */
import {EventEmitter} from "@hyper/event-emitter";
import type {Method} from "@hyper/http";

import createPath from "./utils/createPath.js";
import PathString from "./path/PathString.js";
import PathParams from "./path/PathParams.js";
import PathRegExp from "./path/PathRegExp.js";
import type Request from "./types/Request.js";
import type Response from "./types/Response.js";
import type Handle from "./types/Handle.js";
import type Handler from "./types/Handler.js";
import type HandlerOptions from "./types/HandlerOptions.js";
import type MatchParams from "./path/types/MatchParams.js";
import type Match from "./path/types/Match.js";
import type Path from "./path/types/Path.js";
import type PathSource from "./path/types/PathSource.js";


export interface RouteOptions<Req extends Request = Request, Res extends Response = Response> extends HandlerOptions {
  method: Method;
  path: PathSource;
  handler: Handle<Req, Res>;
}

const defaults = {
  pathOptions: {
    sensitive: false,
    strict: false,
    end: true,
    start: true,
  },
};

type EventMap = {
  error: (error: Error) => void;
  mount: (server: HandlerOptions) => void;
  request: (
    req: Request,
    res: Response,
    path?: string
  ) => void;
};

class Route<Req extends Request = Request, Res extends Response = Response> implements Handler<Req, Res> {
  private parentOptions?: HandlerOptions;
  private options!: RouteOptions<Req, Res>;
  private emitter?: EventEmitter<EventMap>;
  private path!: Path;

  /**
   * Route constructor
   */
  public constructor(options: RouteOptions<Req, Res>) {
    this.setOptions(options);
  }

  /**
   * Get or initialize EventEmitter
   */
  private getEmitter(): EventEmitter<EventMap> {
    return this.emitter ?? (this.emitter = new EventEmitter<EventMap>());
  }
  
  /**
   * Emit event using emitter
   */
  private emit<E extends keyof EventMap>(eventName: E, ...args: Parameters<EventMap[E]>): void {
    if (this.emitter) {
      this
        .emitter
        .emit(eventName, ...args);
    }
  }

  /**
   * Register event handler
   */
  public on<E extends keyof EventMap>(event: E, handler: EventMap[E]): this {
    this
      .getEmitter()
      .on(event, handler);

    return this;
  }

  /**
   * Set path for Route
   */
  public setPath(path: PathSource): this {
    this.path = createPath(path);

    return this;
  }

  /**
   * Get path
   */
  public getPath(): Path {
    return this.path;
  }

  /**
   * Set router options
   */
  public setOptions(options: RouteOptions<Req, Res>): this {
    this.options = options;

    if (this.options.path) {
      this.setPath(this.options.path);
    }

    return this;
  }

  /**
   * Get options
   */
  public getOptions(): RouteOptions<Req, Res> {
    const {
      options,
      parentOptions = {},
    } = this;

    return {
      ...options,
      pathOptions: {
        ...parentOptions.pathOptions,
        ...options.pathOptions,
      },
    };
  }

  /**
   * Set Parent for route
   */
  public setParentOptions(parentOptions: HandlerOptions): this {
    this.parentOptions = parentOptions;

    return this;
  }

  /**
   * Mount parent, freeze parent
   */
  public mount(parentOptions: HandlerOptions): void {
    this.emit("mount", parentOptions);
    this.setParentOptions(parentOptions);

    const path = this.getPath();
    const options = this.getOptions();

    path.mount({
      ...defaults.pathOptions,
      ...options.pathOptions,
    });

    this.selectHandle();
    Object.freeze(this);
  }

  /**
   * Default handle function will throw
   * @throws {Error} if not mounted
   */
  public handler(req: Req, res: Res, path?: string): void {
    this.emit("request", req, res, path);

    throw new Error("Route.handle(..): route is not mounted");
  }

  /**
   * Use Handler for incoming req, res
   * Catch any handler errors
   */
  private async handling(req: Req, res: Res): Promise<void> {
    try {
      await this.options.handler(req, res);
    }
    catch (error: unknown) {
      this.emit("error", error as Error);
      throw error;
    }
  }

  /**
   * Match Route
   */
  private match(req: Request, res: Response, path?: string): Match {
    if (res.writableEnded || this.options.method !== req.method) {
      return false;
    }
    else {
      return this.path.match(path);
    }
  }

  /**
   * Handle incoming request
   * Set Params if matched
   */
  private async handlePathParams(req: Req, res: Res, path?: string): Promise<void> {
    const match = this.match(req, res, path) as MatchParams | false;

    if (match) {
      this.emit("request", req, res);

      req.setRouteParams(match.params);
      req.resetLayerParams();

      await this.handling(req, res);
    }
  }

  /**
   * Handle incoming request
   * Set Params if matched
   */
  private async handlePathRegExp(req: Req, res: Res, path?: string): Promise<void> {
    const match = this.match(req, res, path) as MatchParams | false;

    if (match) {
      this.emit("request", req, res);

      req.setRouteParams(match.params);
      req.resetLayerParams();

      await this.handling(req, res);
    }
  }

  /**
   * Handle incoming request
   */
  private async handlePathString(req: Req, res: Res, path?: string): Promise<void> {
    const match = this.match(req, res, path);

    if (match) {
      this.emit("request", req, res);

      req.resetRouteParams();
      req.resetLayerParams();

      await this.handling(req, res);
    }
  }
  
  /**
   * Set handle function for route
   */
  private selectHandle(): void {
    const path = this.getPath();

    if (path instanceof PathString) {
      this.handler = this.handlePathString;
    }
    if (path instanceof PathParams) {
      this.handler = this.handlePathParams;
    }
    if (path instanceof PathRegExp) {
      this.handler = this.handlePathRegExp;
    }
  }
}

export default Route;
