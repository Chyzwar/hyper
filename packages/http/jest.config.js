
export default {
  displayName: "@hyper/http",
  testEnvironment: "node",
  transform: {},
  testRegex: "(/__tests__/.*|(\\.|/)(test|spec))\\.(js?|jsx?)$",
  roots: ["<rootDir>/lib/"],
};

