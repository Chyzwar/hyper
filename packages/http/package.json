{
  "name": "@hyper/http",
  "version": "0.0.1",
  "engines": {
    "node": "18.x.x",
    "yarn": "1.x.x"
  },
  "type": "module",
  "main": "./lib/index.js",
  "description": "Http client for electron and node",
  "sideEffects": false,
  "scripts": {
    "build": "tsc --build",
    "build:watch": "tsc --build  --watch",
    "lint": "eslint . --ext .ts",
    "lint:fix": "eslint . --fix --ext .ts",
    "test": "jest --passWithNoTests",
    "test:watch": "jest --watch --passWithNoTests"
  },
  "keywords": [
    "http"
  ],
  "author": "chyzwar",
  "license": "UNLICENSED",
  "dependencies": {
    "@hyper/utility-types": "0.x.x"
  },
  "devDependencies": {
    "@chyzwar/eslint-config": "^0.2.14",
    "@chyzwar/tsconfig": "^0.2.14",
    "@swc/core": "^1.3.96",
    "@swc/jest": "^0.2.29",
    "eslint": "^8.53.0",
    "jest": "^29.7.0",
    "typescript": "^5.2.2"
  },
  "files": [
    "src/",
    "lib/",
    "!**/__tests__/",
    "!lib/buildInfo.json"
  ]
}
