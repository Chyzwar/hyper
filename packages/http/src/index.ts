export {default as StatusMessage} from "./StatusMessage.js";
export {default as Localhost} from "./Localhost.js";

export type {default as Body} from "./Body.js";
export {default as Charset} from "./Charset.js";
export {default as ContentEncoding} from "./ContentEncoding.js";
export {default as ContentType} from "./ContentType.js";
export {default as Credentials} from "./Credentials.js";
export {default as getContentType} from "./getContentType.js";
export {default as HeaderName} from "./HeaderName.js";
export type {default as Headers} from "./Headers.js";
export {default as HttpError} from "./HttpError.js";
export {default as Method} from "./Method.js";
export {default as Mode} from "./Mode.js";
export {default as Redirect} from "./Redirect.js";
export {default as StatusCode} from "./StatusCode.js";
export {default as UserAgent} from "./UserAgent.js";






