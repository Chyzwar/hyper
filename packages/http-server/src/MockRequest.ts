/* eslint-disable @typescript-eslint/no-unnecessary-condition */
import {Readable} from "stream";
import {Method, HeaderName, ContentType} from "@hyper/http";
import type {JSONValue} from "@hyper/utility-types";
import type {Headers} from "@hyper/http";
import type {UrlWithParsedQuery} from "url";
import type {ParsedUrlQuery} from "querystring";
import type {Socket} from "net";
import type {ServerHttp2Stream} from "node:http2";

import parseURL from "./utils/parseURL.js";
import type Http1Request from "./request/Http1Request.js";

export type MockRequestEvents = 
  | "aborted" 
  | "close" 
  | "error";

export interface MockRequestOptions {
  method?: Method;
  url?: string;
  data?: object | string;
  headers?: Headers;
  trailers?: Headers;
  simulate?: MockRequestEvents[];
}

class MockRequest<B = unknown> extends Readable implements Http1Request<B> {
  [key: string]: unknown;

  public body!: B;
  public url: string;
  public method: Method;
  public parsedUrl: UrlWithParsedQuery;
  public rawHeaders: string[];
  public headers: Headers;
  public rawTrailers: string[];
  public trailers: Headers;
  public socket!: Socket;
  public connection!: Socket;
  public stream!: ServerHttp2Stream;
  public authority!: string;
  public scheme!: string;
  public routeParams?: Record<string, JSONValue>;
  public layerParams?: Record<string, JSONValue>;
  public routerParams?: Record<string, JSONValue>;
  public httpVersion: string;
  public httpVersionMajor: number;
  public httpVersionMinor: number;
  public complete: boolean;
  public aborted: boolean;
  public data?: object | string;
  public simulate?: string[];
  
  public setTimeout!: (msc: number, callback: () => void) => this;


  public constructor({url = "/", method, headers, trailers, data, simulate}: MockRequestOptions = {}) {
    super();  

    this.httpVersion = "1.1";
    this.httpVersionMajor = 1;
    this.httpVersionMinor = 1;
    this.complete = true;
    this.aborted = false;

    this.data = data;

    this.headers = headers ?? {};
    this.rawHeaders = [];
    this.trailers = trailers ?? {};
    this.rawTrailers = [];
    this.simulate = simulate ?? [];

    this.url = url;
    this.parsedUrl = parseURL(this.url);
    this.method = method ?? Method.GET;
    
    if (data && !this.headers[HeaderName.ContentType]) {
      if (typeof data === "object") {
    
        if (!this.headers[HeaderName.ContentType]) {
          this.headers[HeaderName.ContentType] = ContentType.ApplicationJSON;
        }

        if (!this.headers[HeaderName.ContentLength]) {
          this.headers[HeaderName.ContentLength] = `${Buffer.byteLength(JSON.stringify(data))}`;
        }
      }
    }
  }

  public get query(): ParsedUrlQuery {
    return this.parsedUrl.query;
  }

  public get path(): string {
    return this.parsedUrl.path ?? "/";
  }

  public get params(): object {
    return {
      ...this.layerParams,
      ...this.routerParams,
      ...this.routeParams,
    };
  }

  public _read(): void {
    setImmediate(() => {
      const data = JSON.stringify(this.data);
      this.push(data.slice(0, 1));

      if (this.simulate?.includes("aborted")) {
        this.emit("aborted");
      }
      if (this.simulate?.includes("error")) {
        this.emit("error", new Error("Simulated stream error"));
      }
      if (this.simulate?.includes("close")) {
        this.emit("close");
      }

      this.push(data.slice(1));
      this.push(null); 
    });
  }

  public resetRouterParams(): void {
    this.routerParams = undefined;
  }

  public setRouterParams(params: Record<string, string>): this {
    this.routerParams = params;

    return this;
  }

  public resetRouteParams(): void {
    this.routeParams = undefined;
  }

  public setRouteParams(params: Record<string, string>): this {
    this.routeParams = params;
    
    return this;
  }

  public resetLayerParams(): void {
    this.layerParams = undefined;
  }
}

export default MockRequest;