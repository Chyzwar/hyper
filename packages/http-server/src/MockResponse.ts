import {Writable} from "stream";
import {StatusCode, HeaderName, ContentType, HttpError} from "@hyper/http";
import type {Headers} from "@hyper/http";
import type {Socket} from "net";
import type {ServerHttp2Stream} from "node:http2";

import type Http1Response from "./response/Http1Response.js";
import type ResponseOptions from "./types/ResponseOptions.js";
import type Http1Request from "./request/Http1Request.js";

export type MockResponseEvents = 
  | "error"; 

export interface MockResponseOptions {
  headers?: Headers;
  statusCode?: StatusCode;
  simulate?: MockResponseEvents[];
}

class MockResponse<B = unknown> extends Writable implements Http1Response<B> {
  public options: ResponseOptions = {
    etag: false,
    compression: false,
  };

  public statusCode: StatusCode;
  public headers: Headers;
  public connection!: Socket;
  public socket!: Socket;
  public simulate: string[];
  public statusMessage!: string;
  public shouldKeepAlive!: boolean;
  public chunkedEncoding!: boolean;
  public upgrading!: boolean;
  public finished!: boolean;
  public headersSent!: boolean;
  public sendDate!: boolean;
  public stream!: ServerHttp2Stream;
  public useChunkedEncodingByDefault!: boolean;
  public createPushResponse!: () => void; 
  public assignSocket!: (socket: Socket) => void;
  public detachSocket!: (socket: Socket) => void;
  public addTrailers!: (headers: Headers) => void;
  public writeContinue!: (callback: () => void) => void;
  public writeProcessing!: () => void;
  public flushHeaders!: () => void;
  public getHeaderNames!: () => string[];
  public removeHeader!: (name: string) => void;
  public hasHeader!: (name: string) => boolean;
  public writeEarlyHints!: (hints: Record<string, string[] | string>, callback?: () => void) => void;
  public setTimeout!: (msc: number, callback: () => void) => this;
  public data?: string;
  public req!: Http1Request;

  public constructor({headers, statusCode, simulate}: MockResponseOptions = {}) {
    super();

    this.simulate = simulate ?? [];
    this.headers = headers ?? {};
    this.statusCode = statusCode ?? StatusCode.Ok;
  }

  public getHeaders(): Headers {
    return this.headers;
  }
  
  public getHeader(name: string): string | undefined {
    return this.headers[name];
  }

  public setStatusCode(value: number): this {
    this.statusCode = value;
    return this;
  }
  
  public setHeader(name: string, value: string[] | number | string): this {
    this.headers[name] = `${Array.isArray(value) ? value.join(", ") : value}`;
    return this;
  }
  
  public writeHead(code: StatusCode, ...rest: unknown[]): this {
    this.statusCode = code;

    const [headers]: unknown[] = rest;
    if (headers instanceof Object) {
      Object
        .entries(headers)
        .forEach(([name, value]) => {
          if (typeof value === "string") {
            this.setHeader(name, value);
          }
        });
    }
    return this;
  }

  public _write(chunk: string, _encoding: string, callback: Function): void {
    this.data += chunk;

    if (this.simulate.includes("error")) {
      this.emit("error", new Error("Response simulated error"));
    }
    callback();
  }

  public error(error: Error | HttpError): void {
    if (error instanceof HttpError) {
      this.statusCode = error.statusCode;
      this.statusMessage = error.statusMessage;
    }
    else {
      this.statusCode = StatusCode.InternalServerError;
    }

    if (process.env.NODE_ENV === "development") {
      this.write(`Error: ${String(error)}`);
      this.write(`Stack: ${String(error.stack)}`);  
    }
    this.end();
  }

  public json(object: B): void {
    const body = JSON.stringify(object);
    
    this.writeHead(StatusCode.Ok, {
      [HeaderName.ContentType]: ContentType.ApplicationJSON,
      [HeaderName.ContentLength]: `${Buffer.byteLength(body)}`,
    });
    this.end(body);
  }

  public text(text: string): void {
    this.writeHead(StatusCode.Ok, {
      [HeaderName.ContentType]: ContentType.TextPlain,
      [HeaderName.ContentLength]: `${Buffer.byteLength(text)}`,
    });
    this.end(text);
  }
}

export default MockResponse;