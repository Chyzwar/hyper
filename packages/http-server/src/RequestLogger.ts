/* eslint-disable @typescript-eslint/no-unsafe-enum-comparison */

import {Logger} from "@hyper/logger";
import {HeaderName, StatusCode} from "@hyper/http";

import Layer from "./Layer.js";
import type Request from "./types/Request.js";
import type Response from "./types/Response.js";
import type {LayerOptions} from "./Layer.js";


export interface RequestLoggerOptions extends LayerOptions {
  successMessage?: string;
  errorMessage?: string;
  abortedMessage?: string;
  serializers?: {
    error?: (
      req: Request, 
      res: Response, 
      responseTime: number, 
      error?: Error
    ) => object;
    success?: (
      req: Request, 
      res: Response, 
      responseTime: number
    ) => object;
    abort?: (
      req: Request, 
      res: Response, 
      responseTime: number
    ) => object;
  };
  logger?: {
    info: (message: string, meta: object) => void;
    warn: (message: string, meta: object) => void;
    error: (message: string, meta: object) => void;
  };
}


function baseSerializer(req: Request, res: Response, responseTime: number): object {
  return {
    responseTime,
    req: {
      method: req.method,
      url: req.path,
      query: req.query,
      headers: {
        [HeaderName.UserAgent]: req.headers[HeaderName.UserAgent], 
        [HeaderName.ContentType]: req.headers[HeaderName.ContentType],
      },
    },
    res: {
      statusCode: res.statusCode,
    },
  };
}

const defaults = {
  logger: new Logger(),
  successMessage: "request completed",
  errorMessage: "request error",
  abortedMessage: "request aborted",
  serializers: {
    error: (req: Request, res: Response, responseTime: number, error?: Error): object => {
      const meta = baseSerializer(req, res, responseTime);
      
      if (error) {
        return {
          ...meta,
          error,
        };
      }
      else {
        return meta;
      }
    },
    success: (req: Request, res: Response, responseTime: number): object => {
      const meta = baseSerializer(req, res, responseTime);
      return meta;
    },
    abort: (req: Request, res: Response, responseTime: number): object => {
      const meta = baseSerializer(req, res, responseTime);
      return meta;
    },
  },
};

class RequestLogger extends Layer {
  private readonly successMessage: string;
  private readonly errorMessage: string;
  private readonly abortedMessage: string;

  private readonly logger: {
    info: (message: string, meta: object) => void;
    warn: (message: string, meta: object) => void;
    error: (message: string, meta: object) => void;
  };

  private readonly serializers: {
    success: (req: Request, res: Response, responseTime: number) => object;
    abort: (req: Request, res: Response, responseTime: number) => object;
    error: (req: Request, res: Response, responseTime: number, error?: Error) => object;
  };

  public constructor(options?: RequestLoggerOptions) {
    super(options);
    
    this.successMessage = options?.successMessage ?? defaults.successMessage;
    this.errorMessage = options?.errorMessage ?? defaults.errorMessage;
    this.abortedMessage = options?.abortedMessage ?? defaults.abortedMessage;

    this.logger = options?.logger ?? defaults.logger;
    this.serializers = {
      ...defaults.serializers,
      ...options?.serializers,
    };
  }

  public handler(req: Request, res: Response): void {
    const start = Date.now();

    const onError = (error: Error): void => {
      const end = Date.now();

      this.logger.error(
        this.errorMessage, 
        this.serializers.error(req, res, end - start, error)
      );
    };
  
    const onAborted = (): void => {
      const end = Date.now();

      this.logger.warn(
        this.abortedMessage, 
        this.serializers.abort(req, res, end - start)
      );
    };

    const onFinished = (): void => {
      req.removeListener("error", onError);
      req.removeListener("aborted", onAborted);
      req.removeListener("finish", onFinished);
      
      const end = Date.now();

      if (res.statusCode < StatusCode.BadRequest) {
        this.logger.info(
          this.successMessage, 
          this.serializers.success(req, res, end - start)
        );
      }
      else {
        this.logger.error(
          this.errorMessage, 
          this.serializers.error(req, res, end - start)
        );
      }
    };

    res.on("finish", onFinished);
    res.on("error", onError);
    req.on("aborted", onAborted);
  }
}

export default RequestLogger;