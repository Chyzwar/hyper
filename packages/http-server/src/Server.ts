/* eslint-disable @typescript-eslint/no-misused-promises */

import {EventEmitter} from "@hyper/event-emitter";
import {HttpError, StatusCode} from "@hyper/http";
import {createServer as createHttp1Server} from "http";
import {createServer as createHttp1SecureServer} from "https";
import {
  createServer as createHttp2Server,
  createSecureServer as createHttp2SecureServer,
} from "http2";
import type { 
  RequestListener, 
  Server as Htt1Server,
  ServerOptions as Http1ServerOptions,
} from "http";
import type { 
  Server as Htt1SecureServer,     
  ServerOptions as Http1SecureServerOptions,
} from "https";
import type { 
  Http2Server, 
  Http2SecureServer, 
  Http2ServerRequest, 
  Http2ServerResponse,
  ServerOptions as Http2ServerOptions,
  SecureServerOptions as Http2SecureServerOptions,
} from "http2";
import type {AddressInfo} from "net";

import Http1Response from "./response/Http1Response.js";
import Http2Request from "./request/Http2Request.js";
import Http2Response from "./response/Http2Response.js";
import Http1Request from "./request/Http1Request.js";
import Chain from "./Chain.js";
import MockResponse from "./MockResponse.js";
import MockRequest from "./MockRequest.js";
import parseURL from "./utils/parseURL.js";
import type Request from "./types/Request.js";
import type Response from "./types/Response.js";
import type Router from "./Router.js";
import type Layer from "./Layer.js";
import type Handler from "./types/Handler.js";
import type HandlerOptions from "./types/HandlerOptions.js";
import type ResponseOptions from "./types/ResponseOptions.js";

type NativeOptions = Http1SecureServerOptions & Http1ServerOptions & Http2SecureServerOptions & Http2ServerOptions;

export interface ServerOptions extends NativeOptions {
  port: number;
  http2?: boolean;
  pathOptions?: {
    sensitive?: boolean;
    strict?: boolean;
  };
  requestOptions?: {
    freshness?: boolean;
  };
  responseOptions?: ResponseOptions;
}

enum NativeEventNames {
  checkContinue = "checkContinue",
  request = "request",
  session = "session",
  sessionError = "sessionError",
  stream = "stream",
  timeout = "timeout",
  unknownProtocol = "unknownProtocol",
}

type EventMap = {
  checkContinue: (
    req: Request,
    res: Response,
  ) => void;
  request: (
    req: Request,
    res: Response,
    path?: string
  ) => void;
  chain: (chain: Chain) => void;
  handler: (handler: Handler) => void;
  layer: (layer: Layer) => void;
  router: (router: Router) => void;
  mount: (server: Server) => void;
};

type NativeServer = Htt1SecureServer | Htt1Server | Http2SecureServer | Http2Server;

const defaults: Omit<ServerOptions, "port"> = {
  allowHTTP1: true,
  minVersion: "TLSv1.3",
  IncomingMessage: Http1Request,
  ServerResponse: Http1Response,
  Http1IncomingMessage: Http1Request,
  Http1ServerResponse: Http1Response,
  Http2ServerRequest: Http2Request,
  Http2ServerResponse: Http2Response,
};

class Server implements Handler {
  private readonly options: ServerOptions;
  private readonly server: NativeServer;

  private chain: Chain;
  private emitter?: EventEmitter<EventMap>;

  public constructor(options: ServerOptions) {
    this.options = {
      ...defaults,
      ...options,
    };

    this.chain = new Chain();

    if (this.isSecure()) {
      if (this.options.http2) {
        this.server = createHttp2SecureServer(
          this.options,
          this.handler as unknown as (req: Http2ServerRequest, res: Http2ServerResponse) => void
        );
      }
      else {
        this.server = createHttp1SecureServer(
          this.options,
          this.handler as unknown as RequestListener
        );
      }
    }
    else {
      if (this.options.http2) {
        this.server = createHttp2Server(
          this.options,
          this.handler as unknown as (req: Http2ServerRequest, res: Http2ServerResponse) => void
        );
      }
      else {
        this.server = createHttp1Server(
          this.options,
          this.handler as unknown as RequestListener
        );
      }
    }
  }

  /**
   * Server is secure if key and cert are provided
   */
  public isSecure(): boolean {
    const {
      key,
      cert,
    } = this.options;

    return Buffer.isBuffer(cert) && Buffer.isBuffer(key);
  }

  /**
   * Get instance of underlying server
   */
  public getServer(): NativeServer {
    return this.server;
  }

  /**
   * Get or initialize EventEmitter
   */
  private getEmitter(): EventEmitter<EventMap> {
    return this.emitter ?? (this.emitter = new EventEmitter<EventMap>());
  }

  /**
   * Emit event using emitter
   */
  private emit<E extends keyof EventMap>(eventName: E, ...args: Parameters<EventMap[E]>): void {
    if (this.emitter) {
      this
        .emitter
        .emit(eventName, ...args);
    }
  }
  
  /**
   * Register event handler
   */
  public on<E extends keyof EventMap>(event: E, handler: EventMap[E]): this {
    if (event in NativeEventNames) {
      this
        .getServer()
        .on(event, handler);
    }
    else {
      this
        .getEmitter()
        .on(event, handler);
    }

    return this;
  }

  public getOptions(): ServerOptions {
    return this.options;
  }

  /**
   * Set Application chain
   *
   */
  public setChain(chain: Chain): this {
    this.chain = chain;
    this.emit("chain", chain);

    return this;
  }

  /**
   * Get Application chain
   */
  public getChain(): Chain {
    return this.chain;
  }

  /**
   * Add middleware: Layer
   */
  public addLayer(layer: Layer): this {
    this
      .getChain()
      .add(layer);

    this.emit("layer", layer);

    return this;
  }

  /**
   * Add middleware: Router
   */
  public addRouter(router: Router): this {
    this
      .getChain()
      .add(router);

    this.emit("router", router);

    return this;
  }

  /**
   * Add middleware: Handler
   */
  public add<Req extends Request = Request, Res extends Response = Response>(handler: Handler<Req, Res>): this {
    this
      .getChain()
      .add(handler as Handler);

    this
      .emit("handler", handler as Handler);

    return this;
  }

  public async inject(req: MockRequest = new MockRequest(), res: MockResponse = new MockResponse()): Promise<void> {
    await this.mount(this.getOptions());
    
    return new Promise((resolve, reject) => {
      res.on("close", resolve);

      this.handler(
        req, 
        res
      ).then(() => {
        req.destroy();
        res.destroy();
      }).catch(reject);
    });
  }

  /**
   * Mount application, freeze chain
   */
  public async mount(options: HandlerOptions): Promise<void> {
    if (Object.isFrozen(this.chain)) return;

    this.emit("mount", this);
    const chain = this.getChain();

    await chain.mount(options);
    Object.freeze(chain);
  }

  /**
   * Server request handler, start request pipeline
   */
  public handler: (request: Request, response: Response) => Promise<void> = async(req: Request, res: Response): Promise<void> => {
    try {
      /**
       * Parse url, this cannot be done in constructor because
       * IncomingMessage do not have url parsed in constructor
       */
      req.parsedUrl = parseURL(req.url);

      await this.chain.handler(
        req,
        res,
        req.path
      );
    }

    catch (error: unknown) {
      res.error(error as Error);
    }
    finally {
      if (!res.writableEnded) {
        res.error(new HttpError(StatusCode.NotFound));
      }
    }
  };

  /**
   * Mount and start listening for requests
   */
  public async listen(): Promise<AddressInfo> {
    await this.mount(this.getOptions());
    Object.freeze(this);
    
    return new Promise<AddressInfo>((resolve) => {
      this.server.listen(this.options.port, () => {
        resolve(this.server.address() as AddressInfo); 
      });
    });
  }
}

export default Server;