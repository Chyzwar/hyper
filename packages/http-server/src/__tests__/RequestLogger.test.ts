import {Method, StatusCode} from "@hyper/http";
import {jest} from "@jest/globals";

import RequestLogger from "../RequestLogger.js";
import Route from "../Route.js";
import Server from "../Server.js";
import MockRequest from "../MockRequest.js";
import MockResponse from "../MockResponse.js";


jest.useFakeTimers();
jest.setSystemTime(1);

describe("Request Logger", () => {
  const logger = {
    info: jest.fn(),
    error: jest.fn(),
    warn: jest.fn(),
  };


  const server = new Server({port: 0});
  server.add(new RequestLogger({logger}));

  server.add(new Route({
    path: "/", 
    method: Method.GET,
    handler: (req, res): void => {
      res.json({message: "Info"});
    },
  }));
  server.add(new Route({
    path: "/error", 
    method: Method.GET,
    handler: (): void => {
      throw new Error("Route Error");
    },
  }));

  it("should log success", async() => {
    const req = new MockRequest();
    const res = new MockResponse();
    
    await server.inject(req, res);

    expect(logger.info).toMatchSnapshot();
    expect(res.statusCode).toBe(StatusCode.Ok);
  });

  it("should log route error", async() => {
    const req = new MockRequest({url: "/error"});
    const res = new MockResponse();
    
    await server.inject(req, res);

    expect(logger.error).toMatchSnapshot();
    expect(res.statusCode).toBe(StatusCode.InternalServerError);
  });

  it("should response error", async() => {
    const req = new MockRequest();
    const res = new MockResponse({simulate: ["error"]});

    await server.inject(req, res);

    expect(logger.error).toMatchSnapshot();
  });

  it("should log aborted request", async() => {
    const req = new MockRequest({simulate: ["aborted"]});
    const res = new MockResponse();
    
    await server.inject(req, res);

    expect(logger.warn).toMatchSnapshot();
  });
});