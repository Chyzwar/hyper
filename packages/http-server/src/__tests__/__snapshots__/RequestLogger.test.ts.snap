// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Request Logger should log aborted request 1`] = `[MockFunction]`;

exports[`Request Logger should log route error 1`] = `[MockFunction]`;

exports[`Request Logger should log success 1`] = `[MockFunction]`;

exports[`Request Logger should response error 1`] = `
[MockFunction] {
  "calls": [
    [
      "request error",
      {
        "error": [Error: Response simulated error],
        "req": {
          "headers": {
            "content-type": undefined,
            "user-agent": undefined,
          },
          "method": "GET",
          "query": {},
          "url": "/",
        },
        "res": {
          "statusCode": 200,
        },
        "responseTime": 0,
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
